src/Sarif.Viewer.VisualStudio.ResultSources.DeveloperCanvas.Core/Util.cs

```diff
@@ -0,0 +1,291 @@
+        /// </summary>
+        /// <param name="text">The text to pluralize.</param>
+        /// <param name="count">The number of items that "text" represents.</param>
+        /// <returns>The string with approrpiate plural ending</returns>
+        public static string S(string text, int count)
+        {
+            if (count != 1)
+            {
+                if (text.EndsWith("ch"))
+                {
+                    text += "es";
+                }
+                else
+                {
+                    text += "s";
+                }
+            }
+
+            return $"{count} {text}";
+        }
+
+        /// <summary>
+        /// Returns the code base root along with information about if it is a source depot code base.
+        /// </summary>
+        /// <param name="filePath">File path of a file in the code base</param>
+        /// <param name="rootDir">The root of the code base</param>
+        /// <returns>True if this file is part of a source depot code base.</returns>
+        public static bool IsSourceDepot(string filePath, out string rootDir)
+        {
+            rootDir = null;
+
+            try
+            {
+                rootDir = Path.GetDirectoryName(filePath);
+                while (!string.IsNullOrEmpty(rootDir))
+                {
+                    if (File.Exists(Path.Combine(rootDir, sdIniFileName)))
+                    {
+                        return true;
+                    }
+                    rootDir = Path.GetDirectoryName(rootDir);
+                }
+            }
+            catch (Exception)
+            {
+                // Swallow any exceptions that happen here. This is most likely due to the filePath being invalid.
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Parses the server, project, and repository out of a full git-based repository URL.
+        /// E.g. if <paramref name="gitUrl"/> = "https://dev.azure.com/serverName/projectName/_git/repoName" then
+        /// <paramref name="gitServer"/> = "https://dev.azure.com/serverName",
+        /// <paramref name="projectName"/> = "projectName", and
+        /// <paramref name="gitRepositoryName"/> = "repoName".
+        /// Note that some URLs, like GitHub, may not have a <paramref name="projectName"/>.
+        /// It is the caller's responsibility to check the validity of each extracted part.
+        /// </summary>
+        /// <param name="gitUrl">The git url to be parsed. Ex: https://dev.azure.com/serverName/projectName/_git/repoName</param>
+        /// <param name="gitServer">The server of the repo Ex: dev.azure.com/serverName</param>
+        /// <param name="projectName">The project name of the repo Ex: OS.Fun</param>
+        /// <param name="gitRepositoryName">The name of the repo Ex: devcanvas</param>
+        public static void ParseGitUrl(string gitUrl, out string gitServer, out string projectName, out string gitRepositoryName)
```

Comment:
I had thought that SARIF viewer had some unique but pre-existing functionality to do this.  I wouldn't think we'd want/need to duplicate it? Or in particular have it specific to our plugin?

Reply:
I don't believe it does, the closest thing I can think of is the gitexe functionality it has to read in the git url. I don't think the Sarif Viewer does much specific for ADO repos. @easyrhino-gh is there something that does this sort of functionality already?